### PR TITLE
Adds `validate` package

### PR DESCRIPTION
**Pull request by wcharczuk**
## PR Summary

This PR adds a `validate` package that's meant to mirror the useful parts of JOI. 

It make slight ergonomic tweaks to the `ex` package to allow it to be used for validation errors.

 - **Type:** Features
 - **Intended Change Level:** minor

#### Reviewers:

Reviewers keep the following in mind while reviewing this PR, and provide an assessment of them in your approval comment:

- Does the "Intended Change Level" above match the actual behavior of this PR?
- Is this PR following patterns set forth in the package (if modifying a package), or the go-sdk design patterns if implementing a new package from scratch?
- Does this require a backport to LTS versions? If so ping, let the contributors group know.

### CODE REVIEW

File: validate/string.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+			return ex.New(err)
+		}
+		if !exp.MatchString(string(s)) {
+			return Errorf(ErrStringMatches, "expression: %s", expression)
+		}
+		return nil
+	}
+}
+
+// IsUpper returns a validator if a string is all uppercase.
```
Comment (alee101) [inline validate/string.go]: same for these, either update comment or return a validator?

File: validate/time.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+func (t Time) Before(before time.Time) Validator {
+	return func() error {
+		if time.Time(t).After(before) {
+			return Errorf(ErrTimeBefore, "before: %v", before)
+		}
+		return nil
+	}
+}
+
+// BeforeNowUTC returns a validator that a time should be before a given time.
```
Comment (alee101) [inline validate/time.go]: same as above

File: validate/validated.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package validate
+
+// Validated is a root interface for validated types.
+type Validated interface {
```
Comment (alee101) [inline validate/validated.go]: maybe call this `Validateable`? `IsValidated` suggests there might be a difference between the pre-validation and post-validation state

File: validate/all.go [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package validate
+
+// All returns the all the failing validations.
```
Comment (alee101) [inline validate/all.go]: `returns all the`

File: validate/any.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+func Any(obj interface{}) AnyValidators {
+	return AnyValidators{Obj: obj}
+}
+
+// AnyValidators are validators for the empty interface{}.
+type AnyValidators struct {
+	Obj interface{}
+}
+
+// Zero retruns a validator that asserts an object is it's zero value.
```
Comment (alee101) [inline validate/any.go]: `returns`

File: validate/int.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+		}
+		if int(i) > max {
+			return Errorf(ErrIntMax, "max: %d", max)
+		}
+		return nil
+	}
+}
+
+// Positive returns a validator that an int is positive.
+func (i IntValidators) Positive() error {
```
[the remark this answers is not in the document]
Comment (wcharczuk) [reply validate/int.go]: done

File: validate/string.go (hunk 1 above)
Comment (wcharczuk) [reply validate/string.go]: done

File: validate/time.go (hunk 2 above)
Comment (wcharczuk) [reply validate/time.go]: done

File: validate/validated.go (hunk 3 above)
Comment (wcharczuk) [reply validate/validated.go]: IsValidated is a type assertion that makes sure an object implements a `Validate() error` function, it doesn't connote that function has been called. I should probably make that clear in the docs.

File: validate/all.go (hunk 4 above)
Comment (wcharczuk) [reply validate/all.go]: done

File: validate/any.go (hunk 5 above)
Comment (wcharczuk) [reply validate/any.go]: done

File: validate/int.go [hunk 7]
```diff
@@ -24,7 +24,7 @@ type IntValidators struct {
 func (i IntValidators) Min(min int) Validator {
 	return func() error {
 		if i.Value == nil || *i.Value < min {
-			return Errorf(ErrIntMin, "min: %d", min)
+			return Errorf(ErrIntMin, i.Value, "min: %d", min)
```
Comment (alee101) [inline validate/int.go]: dereference value here and below

Comment (wcharczuk) [reply validate/int.go]: fixed

**Pull request merged**